### PR TITLE
fix(bug): Pass boto_session to SAM Translator library

### PR DESCRIPTION
**Pull request by jfuss**
When validating the sam template, SAM CLI requires credentials to get
the IAM Manged Policies. SAM Translator also requires the region in order
to figure out the parition. Previously, SAM Translator assumed this to be
on the Env but SAM CLI could get this information from a command line argument
or a profile. This commit passes the boto_session into the SAM Translator lib
(v1.35.0 or later), so that SAM Translator can figure out the partition from the
information passed to SAM CLI.

#### Which issue(s) does this change fix?
#442

#### Why is this change necessary?
SAM will through an AttributeError if a region is None (it assumes there will be a region). Locally, there was no way to pass the region/profile into the SAM Translator before v1.35.0. With v1.35.0 and above, there is an optional parameter for passing in a `boto_session` which allows SAM CLI to forward information from the CLI into the SAM Translator Lib.

#### How does it address the issue?
Passes Boto Session to the SAM Translator lib.

#### What side effects does this change have?
None

#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [ ] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by hoffa — APPROVED** [review]

**Review by mndeveci — APPROVED** [review]

File: samcli/commands/validate/validate.py [hunk 1]
```diff
@@ -49,13 +51,20 @@ def do_cli(ctx, template):
...
+    )
 
     try:
         validator.is_valid()
     except InvalidSamDocumentException as e:
         click.secho("Template provided at '{}' was invalid SAM Template.".format(template), bg="red")
         raise InvalidSamTemplateException(str(e)) from e
+    except NoRegionFound as no_region_found_e:
+        raise UserException(
+            "AWS Region was not found. Please configure your region through a profile or --region option",
```
Comment (sriram-mv) [inline samcli/commands/validate/validate.py:65]: should we add AWS_REGION env var too?

**Review by sriram-mv — CHANGES_REQUESTED** [review]

File: samcli/commands/validate/validate.py (hunk 1 above)
Comment (jfuss) [reply samcli/commands/validate/validate.py:65]: I am not sold on adding [`AWS_DEFAULT_REGION`](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#guide-configuration). Yes customers can set this as it is resolved by boto3, but I don't see the need to add every single option here. Most customers are going to use `--region` or define them through a profile locally.

Is there a specific reason why you think we should add it?

Comment (sriram-mv) [reply samcli/commands/validate/validate.py:65]: does sam validate automatically pick up the env vars? just trying to see if we can reduce friction where we can.

**Review by sriram-mv — APPROVED** [review]

File: samcli/commands/validate/validate.py (hunk 1 above)
Comment (jfuss) [reply samcli/commands/validate/validate.py:65]: Yes. We create a `boto_session` like we do in other places throughout. When `None` is passed, boto will resolve for us.

**Pull request merged**